imap/jmap_calendar.c

```diff
@@ -1226,12 +1226,73 @@ static void setcalendar_readalerts(struct jmap_parser *parser,
     *alertsp = alerts;
 }
 
+static void setcalendar_apply_defaultalerts_patch(json_t *arg,
+                                                  const char *userid,
+                                                  const char *mboxname,
+                                                  json_t *invalid)
+{
+    json_t *patches = json_object();  /* Container for defaultAlerts patches */
+    unsigned withTime = 0, withoutTime = 0;
+    const char *field = NULL;
+    json_t *jval;
+
+    json_object_foreach(arg, field, jval) {
+        if (!strncmp(field, "defaultAlertsWithTime/", 22))  {
+            withTime++;
+        }
+        else if (!strncmp(field, "defaultAlertsWithoutTime/", 25))  {
+            withoutTime++;
+        }
+        else {
+            continue;
+        }
+
+        /* Add this patch to container */
+        json_object_set(patches, field, jval);
+    }
+
+    if (withTime || withoutTime) {
+        json_t *cur = json_object();  /* Container for current defaultAlerts */
+
+        if (withTime) {
+            /* Add current defaultAlertsWithTime to container */
+            static const char *annot =
+                DAV_ANNOT_NS "<" XML_NS_CALDAV ">default-alarm-vevent-datetime";
+
+            json_object_set_new(cur, "defaultAlertsWithTime",
+                                getcalendar_defaultalerts(userid, mboxname, annot));
+        }
+        if (withoutTime) {
+            /* Add current defaultAlertsWithoutTime to container */
+            static const char *annot =
+                DAV_ANNOT_NS "<" XML_NS_CALDAV ">default-alarm-vevent-date";
+
+            json_object_set_new(cur, "defaultAlertsWithoutTime",
+                                getcalendar_defaultalerts(userid, mboxname, annot));
+        }
+
+        json_t *new = jmap_patchobject_apply(cur, patches, invalid);
+
+        /* Add patched defaultAlerts to /set args */
+        json_object_foreach(new, field, jval) {
+            json_object_set(arg, field, jval);
```

Comment:
jval could be an empty JSON object. In that case, do set `json_null()`, rather than an empty object.

Reply:
Fixed